### PR TITLE
dev-ui & contract reads & writes

### PR DESCRIPTION
**Pull request by soundyogi**
tested:
- swap stETH (and stake it)
- swap stETH (and not stake it)
- stake rETH (and gain per-block rewards)
- unstake rETH
- claim rETH staking emission

missing (next PR):
- admin functions (wightdraw, mintRocketEther
- update state after transaction has finished

image

### CODE REVIEW

**Comment by soundyogi** [comment]
image

**Comment by soundyogi** [comment]
- swap stETH (and stake it)
  - swap stETH (and not stake it)
  - stake rETH (and gain per-block rewards)
  - unstake rETH
  - claim rETH staking emissions

**Comment by soundyogi** [comment]
image

**Comment by xrchz** [comment]
I wonder if we could do the abi without checking in the file, and just refer to the build artifact (possibly symlink) as something to fetch? Nbd if you prefer not - just to avoid having to keep two files in sync manually.

File: frontend/src/components.mjs [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+      return this.render(balance)
+    })
+
+    this.render(); 
+  }
+  attributeChangedCallback() { this.render(); }
+});
+
+
+// wait until a state key is not undefined, then disaplay that value
```
Comment (xrchz) [inline frontend/src/components.mjs:174]: typo

**Review by xrchz — APPROVED** [review]

File: frontend/src/components.mjs (hunk 1 above)
Comment (soundyogi) [reply frontend/src/components.mjs]: good catch, thanks - got it.

**Comment by soundyogi** [comment]
> I wonder if we could do the abi without checking in the file, and just refer to the build artifact (possibly symlink) as something to fetch? Nbd if you prefer not - just to avoid having to keep two files in sync manually.
> 
> EDIT: maybe your way is better actually, so everything is easy to serve after clone.

tbh. I tried to get ape working but it fails with starting anvil - so I got lazy.
but yeah - it also "easier" - we have to copy the "dist" version of ethers anyway so I dont mind copy/pasting.

I agree a single source of truth should be used - so lets commit the last generated ABI to git? why not?

we could even shave off bytes by converting it to human readable format - but we optimize at the end hopefully :D

**Pull request merged**